Add logic to only update specific categories

schroda opened the pull request:
Makes it possible to only update specific categories.

In case a manga is in an excluded category it will be excluded even if it is also in an included category.

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/controller/UpdateController.kt:
```diff
@@ -93,11 +93,14 @@ object UpdateController {
             .distinctBy { it.id }
             .sortedWith(compareBy(String.CASE_INSENSITIVE_ORDER, MangaDataClass::title))
-            .filter { it.updateStrategy == UpdateStrategy.ALWAYS_UPDATE }
+            .filter { it.updateStrategy == UpdateStrategy.ALWAYS_UPDATE && !categoriesToExclude.any { category -> CategoryManga.getMangaCategories(it.id).contains(category) } }
```
This will create a massive amount of queries to the database, there would probably be a better way.

schroda replied:
good point, didn't think about that

changed it to get the categories for all mangas of the categories that should get updated in one query (not sure if that can be optimized, my sql knowledge is rotting away and it took me way too long to get it working with exposed 😭)

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/controller/UpdateController.kt:
```diff
@@ -93,11 +93,20 @@ object UpdateController {
+        val mangasToCategoriesMap = CategoryManga.getMangasCategories(unfilteredMangasToUpdate.map { it.id })
+        val mangasToUpdate = unfilteredMangasToUpdate.filter {
+            it.updateStrategy == UpdateStrategy.ALWAYS_UPDATE &&
+                !categoriesToExclude.any { category -> mangasToCategoriesMap[it.id]!!.contains(category) }
```
No need to enforce not null
```suggestion
                !categoriesToExclude.any { category -> mangasToCategoriesMap[it.id]?.contains(category) == true }
```

schroda committed:
Add logic to only update specific categories

Makes it possible to only update specific categories.

In case a manga is in an excluded category it will be excluded even if it is also in an included category.

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/controller/UpdateController.kt:
```diff
@@ -93,11 +90,29 @@ object UpdateController {
+                val isMangaInExcludedCategory = excludedCategories.any { category -> mangasToCategoriesMap[it.id]?.contains(category) == true } ?: false
+
+                alwaysUpdate && !isMangaInExcludedCategory
+            }
```
```suggestion
        val mangasToUpdate = categoriesToUpdateMangas
            .filter { it.updateStrategy == UpdateStrategy.ALWAYS_UPDATE }
            .filter {
                !excludedCategories.any { category -> mangasToCategoriesMap[it.id]?.contains(category) == true }
            }
```

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/model/dataclass/CategoryDataClass.kt:
```diff
@@ -1,16 +1,27 @@
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+enum class IncludeInUpdate(@JsonValue val value: Int) {
+    EXCLUDE(0), INCLUDE(1), UNSET(2);
```
Why is UNSET 2 and not 0? I image 0 as a sane default

schroda replied:
no real reason, I just went with boolean (0 = false/exclude, 1 = true/include) which left 2 for UNSET

AriaMoradi replied:
well if we are mapping to numbers, -1 would be better for UNSET as it's not a natural number.

AriaMoradi replied:
But all things considered, as this is not c/c++ it doesn't matter which number we are mapping to.

schroda replied:
I also thought about -1, I prefer that over having UNSET as 0

schroda replied:
I changed it to -1

Syer10 commented on server/src/main/kotlin/suwayomi/tachidesk/manga/impl/CategoryManga.kt:
```diff
@@ -116,4 +116,21 @@ object CategoryManga {
+        }
+
+        return mangasToCategoriesMap
+    }
```
```suggestion
    fun getMangasCategories(mangaIDs: List<Int>): Map<Int, List<CategoryDataClass>> {
        return buildMap {
            transaction {
                CategoryMangaTable.innerJoin(CategoryTable)
                    .select { CategoryMangaTable.manga inList mangaIDs }
                    .groupBy { it[CategoryMangaTable.manga] }
                    .forEach {
                        val mangaId = it.key.value
                        val categories = it.value

                        set(mangaId, categories.map { category -> CategoryTable.toDataClass(category) })
                    }
            }
        }
    }
```

AriaMoradi commented on server/src/main/kotlin/suwayomi/tachidesk/manga/controller/UpdateController.kt:
```diff
@@ -93,11 +90,25 @@ object UpdateController {
+        val unsetCategories = includeInUpdateStatusToCategoryMap[IncludeInUpdate.UNSET].orEmpty()
+        val categoriesToUpdate = includedCategories.ifEmpty { unsetCategories }
+
+        logger.info { "Updating categories: '${categoriesToUpdate.joinToString("', '") { it.name }}'" }
```
make this logger.debug
